### DISCUSSION TITLE
[Share] Set Better Notes Syncing (menu/on closing item/on creating note)

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#140] [Action Scripts]
### Description

Set Better Notes syncing by:
 - menu or shortcut - set `Event` to `None`
 - on closing item (finish reading a paper and sync its child notes) - set `Event` to `Close Tab`
 - on creating note (auto-sync every new note) - set `Event` to `Create Item`

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Set Better Notes Syncing (menu/on closing item/on creating note)
 * @author windingwind
 * @usage See link
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/140/
 * @see https://github.com/windingwind/zotero-better-notes/
 */

if (!item) {
	return;
}

let toSync = [];
if (item.isNote()) {
	toSync.push(item);
}
else if (item.isRegularItem()) {
	toSync.push(...Zotero.Items.get(item.getNotes()));
}

// Filter notes by collection name, so that only notes in this collection will be auto-synced
const FILTER_BY_COLLECTION_NAME = null;

toSync = toSync.filter(noteItem => !Zotero.BetterNotes.api.sync.isSyncNote(noteItem.id))
	.filter(noteItem =>
		!FILTER_BY_COLLECTION_NAME
		|| Zotero.Collections.get(noteItem.getCollections())
			.map(collection => collection.name)
			.includes(FILTER_BY_COLLECTION_NAME));

if (toSync.length === 0) {
	return;
}

const window = require("window");
const PathUtils = require("PathUtils");
const FilePicker = window.require('zotero/modules/filePicker').default;

// Get sync directory

let rootDir = Zotero.Prefs.get("extensions.actionsTags.BNSyncRootDir", true);
if (!rootDir) {
	const fp = new FilePicker();
	fp.init(window, "Choose syncing root directory", fp.modeGetFolder);
	fp.appendFilters(fp.filterAll);

	const rv = await fp.show();
	if (rv !== fp.returnOK && rv !== fp.returnReplace) {
		return;
	}

	rootDir = fp.file;
	Zotero.Prefs.set("extensions.actionsTags.BNSyncRootDir", rootDir, true);
}

await Zotero.BetterNotes.api.$export.syncMDBatch(rootDir, toSync.map(noteItem => noteItem.id));

return `[Sync Notes] Synced ${toSync.length} note items.`;
```

**Comment by cs123951** [comment]
`const FilePicker = window.require('zotero/modules/filePicker').default;` may cause error.
I changed to 
`var { FilePicker } = ChromeUtils.importESModule('chrome://zotero/content/modules/filePicker.mjs');`.
It works.

**Reply by ulahcherubim** [reply]
I happened to have the error you mentioned but after I updated my script with your line, I noticed the notes are not syncing to the directory that I want. How can I choose the directory, do you know?

**Reply by cs123951** [reply]
Sorry, I am using FilePicker for another usage. I did not use other codes in this script.

**Reply by Asstrole** [reply]
> I happened to have the error you mentioned but after I updated my script with your line, I noticed the notes are not syncing to the directory that I want. How can I choose the directory, do you know?

I had same error and used his fix, but had similar problems you had since the script remembers the first directory you chose.

However, I commented out this line and it now asks for the directory every time (which is what I needed).
`Zotero.Prefs.set("extensions.actionsTags.BNSyncRootDir", rootDir, true);`

**Reply by HustleLiam** [reply]
I encountered such an error message: ScriptError:Modulezotero/modules/filePicker is not found at chrome://zotero/content/modules/filePicker.js. Could this be because this module was modified in Zotero 7?

**Reply by HustleLiam** [reply]
> I encountered such an error message: ScriptError:Modulezotero/modules/filePicker is not found at chrome://zotero/content/modules/filePicker.js. Could this be because this module was modified in Zotero 7?

After replacing the code you provided, the problem was solved. Ignore my stupid question.

**Comment by DH-7488** [comment]
I'm trying to sync notes from one Zotero library to two different folders in a single Obsidian vault. I keep seeing the error "Sync Error: TypeError: changes in undefined". Any idea what this means?

**Reply by windingwind** [reply]
Syncing is one-to-one. It’s not possible to sync to two positions

**Reply by DH-7488** [reply]
Thanks so much for your reply. I've been receiving an error message: "Sync error: TypeError changes is undefined". I assumed it had to do with this. 

**UPDATE: I used the debug log to discover syncing was hung up on a corrupt note. Fixed now. Thanks for creating this awesome tool! **

**Reply by Asstrole** [reply]
If you still need sync to multiple locations, this will do what you want https://freefilesync.org I've used it for about 10 years to sync enormous amounts of files/folders. Great program GL

**Comment by 3zra47** [comment]
I tried to get this to work, but it doesn't do anything for me—I don't even get an error. I think I did everything right so I'm not sure what the issue is.

**Reply by windingwind** [reply]
Pls share screenshots

**Reply by 3zra47** [reply]
https://github.com/windingwind/zotero-actions-tags/discussions/140#discussioncomment-15008093

**Comment by 3zra47** [comment]
Here are the screenshots. In the data field I just pasted what I copied up above.

Also, I'm a bit confused about what the purpose of auto-sync in the Better Notes settings are. It doesn't seem to sync my markdown notes which I use in obsidian, so I manually sync it using the sync now button. 

<img width="375" height="487" alt="Screenshot 2025-11-18 at 8 06 50 PM" src="https://github.com/user-attachments/assets/35abf486-e11b-43d4-a477-d153201c9ef4" />
<img width="376" height="487" alt="Screenshot 2025-11-18 at 8 07 02 PM" src="https://github.com/user-attachments/assets/2eaa5586-f629-4bf9-b95a-ffefee6b6952" />
<img width="375" height="486" alt="Screenshot 2025-11-18 at 8 07 15 PM" src="https://github.com/user-attachments/assets/7d28e232-e778-46ca-bd70-3c85a859fcf1" />

**Reply by windingwind** [reply]
> Also, I'm a bit confused about what the purpose of auto-sync in the Better Notes settings are. It doesn't seem to sync my markdown notes which I use in obsidian

It's syncing your notes to markdown files (and back). It won't magically detect what you have outside Zotero and magically turn them into synced notes. You'll need to use import from markdown to get them managed by BN.

Let's first make sure the syncing works, then check the issue you mentioned here.

**Reply by 3zra47** [reply]
>It's syncing your notes to markdown files (and back). It won't magically detect what you have outside Zotero and magically turn them into synced notes. You'll need to use import from markdown to get them managed by BN.

@windingwind Yeah I understand that—I was referring to the Zotero better notes that I then turned into markdown files that are synced in my obsidian vault. I meant that once I sync them with obsidian I have to manually sync the notes—I can't figure out how to get auto sync to work. 

>Let's first make sure the syncing works, then check the issue you mentioned here.

The syncing itself works, but the auto-syncing doesn't seem to. What can I do to try to get that to work? Thank you.

**Reply by 3zra47** [reply]
@windingwind Any ideas on this?

**Reply by windingwind** [reply]
- Do the synced notes appear in the sync manager (under Tools menu), and does syncing them manually again from the sync manager work?
- If not, try to reset the builtin note templates and try again.
- If not, what paths do your synced markdown files have? Are they on special directory, e.g. WSL, remote disk, etc?

**Reply by 3zra47** [reply]
@windingwind the synced notes do appear in the sync manager (under Tools menu), and syncing them manually again from the sync manager does work. It's now auto-syncing, but only after I've left the tab where the better note is open and only after I don't have the connected article selected in my library. Is that how it is supposed to work or is it supposed to auto-sync even when you still have the tab with the better note open or the related note selected in the library? Either way this is a big improvement over having to manually sync.

**Reply by windingwind** [reply]
yes. because it's not safe to sync the note while people are editing it. the sync will happen once you close the note editor.

**Comment by xfap** [comment]
@windingwind Hi, Is there a script or API to trigger a bi-direction sync, I want to trigger a instant sync for a note by shortcut. Thanks!

**Reply by windingwind** [reply]
`Zotero.BetterNotes.hooks.onSyncing()`

**Comment by docsez** [comment]
Is there a way to only sync the template note created by Better Notes and ignore other random notes attached to the item?
